datafusion/physical-plan/src/sorts/merge.rs

```diff
@@ -156,12 +164,22 @@ impl<C: CursorValues> SortPreservingMergeStream<C> {
         }
         // try to initialize the loser tree
         if self.loser_tree.is_empty() {
-            // Ensure all non-exhausted streams have a cursor from which
-            // rows can be pulled
-            for i in 0..self.streams.partitions() {
-                if let Err(e) = ready!(self.maybe_poll_stream(cx, i)) {
-                    self.aborted = true;
-                    return Poll::Ready(Some(Err(e)));
+            // Ensure all non-exhausted streams have a cursor from which rows can be pulled
```

Comment:
This comment implies to me that the code would / should `poll` *all* the streams. However, the code seems to ensure now that only streams that had previously not returned `Ready` for a poll are now polled.

Reply:
IMO, the behavior is more correct now. In the previous version, let's assume the 1st partition is exhausted and returns `None` without setting its cursor. Then, the 2nd partition returns `Pending`. When `poll_next_inner()` is polled again, the iteration starts from the 1st partition, which has already returned None AFAIK polling exhausted streams could cause problems). Therefore, I track which streams have returned a result (either `None` or `Some()`), and which ones have returned `Pending` only.

Reply:
see https://github.com/synnada-ai/datafusion-upstream/pull/34 for alternate idea

Reply:
I've tried to explain my concern with that: https://github.com/synnada-ai/datafusion-upstream/pull/34#issuecomment-2330825156

Reply:
👍  -- response in https://github.com/synnada-ai/datafusion-upstream/pull/34#issuecomment-2331231064